### ISSUE TITLE
onErrorDropped on error signal with windowUntil(…).flatMap(…)/windowUntil(…).concatMap(…) sequence

### ISSUE THREAD

**Issue by mp911de** [reactor/reactor-core#2259]
## Expected Behavior
Error signal should be propagated through the sequence

## Actual Behavior
Error signal propagated followed by `Operator called default onErrorDropped`

## Steps to Reproduce

```java
EmitterProcessor<String> source = EmitterProcessor.create();

source.windowUntil(String::isEmpty)
    .concatMap(Function.identity())
    .subscribe(o -> System.out.println("Result object " + o), Throwable::printStackTrace);

source.onNext("hello");
source.onNext("world");
source.onError(new IllegalStateException());
```

## Possible Solution
Not sure. 

## More Context

The issue popped up initially in an R2DBC driver (r2dbc/r2dbc-mssql#162) where a processor is used to consume incoming Reactor Netty messages. Upon completion, the driver turns a completion signal into an error. The heavily simplified code above is able to reproduce the error. A different reproducer variant is (which is closer to what the R2DBC driver does):

``` java
EmitterProcessor<String> source = EmitterProcessor.create();

source.windowUntil(String::isEmpty)
    .flatMap(it -> it)
    .doOnNext(it -> {

        if (it.equals("fail")) {
            source.onError(new IllegalStateException("fail"));
        }

    }).subscribe(o -> System.out.println("Result object " + o), Throwable::printStackTrace);
;

source.onNext("hello");
source.onNext("world");
source.onNext("fail");
source.onComplete();
```

* Reactor version(s) used: 3.3.7.RELEASE, 3.4.0-M1

**Comment by simonbasle** [comment]
Is the `sink()` used on the processor in real code? I don't think it is the root cause but it would be cleaner to use it anyway.

The problem I see with the second approach at least is that it reinjects the `error` signal back into the `source`, upstream of the windowing. This is bound to compete with `onComplete()`, so ideally you'd need to either introduce mutual exclusion between `source.onComplete()` and `source.onError` calls, or turn the problem around.

Why not throw in the `doOnNext` (or use `handle`) ? This way the `source` is predictable and deterministic, and the `"fail"` string only affects the downstream operator(s).

**Comment by simonbasle** [comment]
Also note that an error in the source of a window operator is currently propagated to both the main sequence and each individual open windows, which is probably why `concatMap` ends up thinking it received 2 onError signals in the first example.

**Comment by mp911de** [comment]
The actual style, how the error gets sent didn't make a difference on `onErrorDropped`. In the SQL Server R2DBC driver, we want to change the completion signal into an error signal. Likely, an operator would be best.

**Comment by simonbasle** [comment]
Really any solution that triggers the error _after_ the `window` should work. `.concatMap(it -> it).concatWith(Mono.error(...))` would work for instance.

Alternatively, when consuming the windows (ie in the concatMap), one could use an `onError` recovery operator on the inner windows themselves.

I think windowing operators absolutely _NEED_ to propagate an upstream error to both their open windows and their downstream, in order to guarantee termination and guarantee that _somebody_ sees the error no matter the way windows are consumed.

**Comment by mp911de** [comment]
The `onError` operator within each window could be a workaround, for now, to prevent bubbling the exception upstream. One thing makes me wonder, why even with a single open window (as per expectation), there seem to be multiple open windows.

**Comment by simonbasle** [comment]
(note that in 3.4 `onErrorDropped` will stop bubbling the exception by default and only log it)

**Comment by simonbasle** [comment]
it is more a factor of the unique window possibly being consumed in a way that doesn't cause the main sequence to terminate. eg. if there is an error handling operator or `materialize()` or the like, the main sequence consuming the window (aka flatmap or concatmap) would only see an `onNext`. So on top of notifying the window, we need to notify the downstream operator itself.

Closing this for now: I think the workaround and the fix for #1431 in 3.4.0 will make this a bit more palatable.
